test/plugin/derivation_endpoint_test.rb

```diff
@@ -0,0 +1,1113 @@
+        assert_match "attachment; ", response[1]["Content-Disposition"]
+
+        response = @uploaded_file.derivation_response(:gray, env: {}, disposition: "inline")
+        assert_match "inline; ", response[1]["Content-Disposition"]
+
+        response = @uploaded_file.derivation_response(:gray, env: {}, disposition: -> { "inline" })
+        assert_match "inline; ", response[1]["Content-Disposition"]
+      end
+
+      it "applies :filename" do
+        @shrine.plugin :derivation_endpoint, filename: "one"
+        response = @uploaded_file.derivation_response(:gray, env: {})
+        assert_match "inline; filename=\"one\"", response[1]["Content-Disposition"]
+
+        @shrine.plugin :derivation_endpoint, filename: -> { "one" }
+        response = @uploaded_file.derivation_response(:gray, env: {})
+        assert_match "inline; filename=\"one\"", response[1]["Content-Disposition"]
+
+        response = @uploaded_file.derivation_response(:gray, env: {}, filename: "two")
+        assert_match "inline; filename=\"two\"", response[1]["Content-Disposition"]
+
+        response = @uploaded_file.derivation_response(:gray, env: {}, filename: -> { "two" })
+        assert_match "inline; filename=\"two\"", response[1]["Content-Disposition"]
+      end
+
+      it "handles Range requests" do
+        response = @uploaded_file.derivation_response(:gray, env: {
+          "HTTP_RANGE" => "bytes=0-3",
+        })
+
+        assert_equal 206,            response[0]
+        assert_equal "bytes 0-3/12", response[1]["Content-Range"]
+        assert_equal "bytes",        response[1]["Accept-Ranges"]
+        assert_equal "gray",         response[2].enum_for(:each).to_a.join
+      end
+
+      it "succeeds when derivative is a File object" do
+        path = Dir::Tmpname.create("") {}
+        @shrine.derivation(:gray) do |file|
+          file = File.open(path, "w+")
+          file << "file content"
+          file
+        end
+        response = @uploaded_file.derivation_response(:gray, env: {})
+
+        assert_equal "file content", response[2].enum_for(:each).to_a.join
+
+        response[2].close
+        refute File.exist?(path)
+      end
+    end
+
+    describe "with upload" do
+      before do
+        @shrine.plugin :derivation_endpoint, upload: true
+      end
+
+      it "returns local file response the first time" do
+        response = @uploaded_file.derivation_response(:gray, env: {})
+
+        assert_equal 200, response[0]
+
+        content_disposition = ContentDisposition.inline("gray-#{@uploaded_file.id}")
+
+        assert_equal "12",                       response[1]["Content-Length"]
+        assert_equal content_disposition,        response[1]["Content-Disposition"]
+        assert_equal "application/octet-stream", response[1]["Content-Type"]
+        assert_equal "bytes",                    response[1]["Accept-Ranges"]
+
+        assert_equal "gray content", response[2].enum_for(:each).to_a.join
```

Comment:
Not seeing how this is confirming it's local file and similar for test below, what's the test that says it fetched in from storage and not create a new derivative.

Reply:
You're right, we should test that the uploaded derivative is not downloaded.